finality-aleph/src/sync/substrate/status_notifier.rs

```diff
@@ -59,19 +104,46 @@ where
     B: BlockT,
     B::Header: SubstrateHeader<Number = BlockNumber>,
 {
-    type Error = Error;
+    type Error = Error<B>;
 
     async fn next(&mut self) -> Result<ChainStatusNotification<B::Header>, Self::Error> {
+        if self.catching_up {
+            match self
+                .header_at(self.last_reported + 1)
+                .map_err(Error::ChainStatus)?
+            {
+                Some(header) => {
+                    self.last_reported += 1;
+                    return Ok(ChainStatusNotification::BlockImported(header));
+                }
+                None => {
+                    self.catching_up = false;
+                    self.trying_since = Instant::now();
+                }
+            }
+        }
         select! {
             maybe_block = self.finality_notifications.next() => {
+                self.trying_since = Instant::now();
                 maybe_block
                     .map(|block| ChainStatusNotification::BlockFinalized(block.header))
                     .ok_or(Error::JustificationStreamClosed)
             },
             maybe_block = self.import_notifications.next() => {
+                if let Some(block) = &maybe_block {
+                    let number = block.header.id().number();
+                    if number > self.last_reported {
+                        self.last_reported = number;
+                    }
+                }
+                self.trying_since = Instant::now();
                 maybe_block
                 .map(|block| ChainStatusNotification::BlockImported(block.header))
                 .ok_or(Error::ImportStreamClosed)
+            },
+            _ = sleep(Duration::from_secs(2).saturating_sub(Instant::now() - self.trying_since)) => {
```

Comment:
I think `Instant::now() - self.trying_since` can panic. Better to use something which has `try` or `saturating` in the name.

Reply:
I don't see any unsafe subtraction in the current version so I guess it cannot panic anymore :) 

You can resolve the comment after you read it.

Reply:
Oh wait, I'm not sure what happens if you sleep `0`. Does it panic?

Reply:
Nope, [the documentation](https://docs.rs/tokio/latest/tokio/time/fn.sleep.html#panics) doesn't mention it as a possibility when panicking.